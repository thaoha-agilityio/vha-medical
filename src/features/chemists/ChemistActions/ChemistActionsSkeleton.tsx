import { Button, InputSearch, Select } from '@/components/ui';

// Types
import { ROLE } from '@/types';

// Config
import { auth } from '@/config/auth';

export const ChemistActionsSkeleton = async () => {
  const { role = ROLE.NORMAL_USER } = (await auth())?.user || {};
  const isAdmin = role === ROLE.ADMIN;

  return (
    <div className="flex flex-col mt-3 md:flex-row gap-4 md:mb-8">
      <InputSearch placeholder="Search Appointments" isDisabled={true} />
      <div className="flex justify-between md:gap-4 mb-10 md:mb-0 ">
        <Select
          aria-label="Select Specialty"
          options={[]}
          isDisabled={true}
          placeholder="Specialty"
          classNames={{
            innerWrapper: 'w-[180px]',
            trigger: 'w-[180px] h-[52px]',
            listbox: 'px-0',
          }}
        />
        {isAdmin && (
          <Button className="h-[52px] font-medium" isDisabled={true}>
            Create
          </Button>
        )}
      </div>
    </div>
  );
};
