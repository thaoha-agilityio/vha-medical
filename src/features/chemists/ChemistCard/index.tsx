'use client';
import { lazy, useCallback, useState } from 'react';

// Components
import {
  Card,
  CardBody,
  CardFooter,
  CardHeader,
  useDisclosure,
} from '@nextui-org/react';
import { Avatar, Button, Text } from '@/components/ui';
import { DeleteIcon, EditIcon, NoteIcon, StarIcon } from '@/icons';
const ConfirmModal = lazy(() => import('@/components/ui/ConfirmModal'));

// Types
import { STATUS_TYPE, UserModel } from '@/types';
import { AVATAR_THUMBNAIL, ERROR_MESSAGE, SUCCESS_MESSAGE } from '@/constants';

// Services
import {
  updateUnpublishAppointment,
  updateUnpublishNotification,
  updateUnpublishUser,
} from '@/services';

// Hooks
import { useToast } from '@/context/toast';

// Utils
import { cn, formatNumberWithUnit } from '@/utils';

export interface ChemistCardProps {
  id: string;
  data: UserModel;
  isAdmin: boolean;
  onEdit?: ({ data, id }: { data: UserModel; id: string }) => void;
}

const ChemistCard = ({ id, data, isAdmin, onEdit }: ChemistCardProps) => {
  const {
    username = '',
    description = '',
    rating = 0,
    reviews = 0,
    tasks = 0,
    specialtyId,
    avatar = AVATAR_THUMBNAIL,
  } = data || {};

  const { data: dataSpecialty } = specialtyId || {};
  const { attributes: attributesSpecialty } = dataSpecialty || {};
  const { name: specialty = '' } = attributesSpecialty || {};

  const openToast = useToast();

  const {
    isOpen: isOpenConfirm,
    onClose: onClosConfirm,
    onOpen: onOpenConfirm,
  } = useDisclosure();

  const [isLoading, setIsLoading] = useState(false);

  const handleDelete = useCallback(async () => {
    setIsLoading(true);
    const { error: errorUser } = await updateUnpublishUser(id);
    if (errorUser) {
      openToast({
        message: ERROR_MESSAGE.DELETE('chemist'),
        type: STATUS_TYPE.ERROR,
      });
      setIsLoading(false);
      return;
    }

    openToast({
      message: SUCCESS_MESSAGE.DELETE('chemist'),
      type: STATUS_TYPE.SUCCESS,
    });

    setIsLoading(false);
    onClosConfirm();

    await updateUnpublishAppointment(id);
    await updateUnpublishNotification(id);
  }, [id, onClosConfirm, openToast]);

  const handleEdit = useCallback(() => {
    onEdit?.({ data, id });
  }, [data, id, onEdit]);

  return (
    <>
      <div className="min-w-[300px] w-full h-[232px] relative">
        <Card
          as="article"
          aria-label="chemist-card"
          className={cn(
            'bg-background-200 w-full h-full p-6 gap-6 overflow-visible',
            isAdmin ? 'pt-0' : '',
          )}
        >
          {isAdmin && (
            <div className="group">
              <div className="absolute z-20 inset-0 rounded-large opacity-0 top-0 left-0 right-0 bottom-0 bg-primary-200 dark:group-hover:opacity-50 group-hover:opacity-70 flex justify-center items-center">
                <Button isIconOnly onClick={handleEdit} aria-label="edit-btn">
                  <EditIcon customClass="text-background-100 flex-shrink-0 w-4 h-4" />
                </Button>
                <Button
                  isIconOnly
                  onClick={onOpenConfirm}
                  aria-label="delete-btn"
                >
                  <DeleteIcon customClass="text-background-100 flex-shrink-0 w-4 h-4" />
                </Button>
              </div>
            </div>
          )}
          <CardHeader className="flex justify-between p-0">
            <div className="flex items-center gap-2">
              <Avatar src={avatar} size="lg" className="aspect-square" />
              <div className="flex flex-col gap-1">
                <Text size="md" variant="title">
                  {username}
                </Text>
                {specialty && (
                  <Text
                    size="xs"
                    variant="description"
                    customClass="font-normal"
                  >
                    {specialty}
                  </Text>
                )}
              </div>
            </div>
          </CardHeader>
          <CardBody className="p-0">
            <Text
              size="sm"
              type="wrap"
              variant="description"
              customClass="line-clamp-2"
            >
              {description || 'No description'}
            </Text>
          </CardBody>
          <CardFooter className="p-0 inline-block rounded-none">
            <div className="flex flex-col gap-1">
              <div className="flex justify-between">
                <div className="flex items-center gap-1 sm:gap-2">
                  <NoteIcon customClass="w-6 h-6" />
                  <Text size="sm" variant="title" customClass="font-medium">
                    {formatNumberWithUnit(tasks, 'Task')}
                  </Text>
                </div>

                <div className="flex items-center gap-1 sm:gap-2">
                  <Text size="sm" variant="title" customClass="font-medium">
                    ({formatNumberWithUnit(reviews, 'Review')})
                  </Text>
                </div>
              </div>

              <div className="flex justify-between items-center gap-1 sm:gap-2">
                <Text size="sm" variant="title" customClass="font-medium">
                  Rating: {rating}
                </Text>
                <div className="flex">
                  {new Array(rating).fill(0).map((_, index) => (
                    <StarIcon
                      key={index}
                      customClass="w-6 h-6 text-light-orange"
                    />
                  ))}
                </div>
              </div>
            </div>
          </CardFooter>
        </Card>
      </div>
      <ConfirmModal
        title="Confirmation"
        subTitle="Do you want to delete this chemist?"
        isOpen={isOpenConfirm}
        isLoading={isLoading}
        onClose={onClosConfirm}
        onAction={handleDelete}
      />
    </>
  );
};

export default ChemistCard;
