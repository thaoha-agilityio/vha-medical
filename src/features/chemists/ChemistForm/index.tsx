'use client';

import {
  ChangeEvent,
  memo,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react';
import { Controller, SubmitHandler, useForm } from 'react-hook-form';

// Components
import { Button, ImageUpload, Input, Select, Text } from '@/components/ui';
import { EyeIcon, EyeSlashIcon, NoteIcon, StarIcon } from '@/icons';
import { Textarea } from '@nextui-org/react';

// Types
import {
  UserModel,
  ChemistFormData,
  UserPayload,
  RolePermission,
  ROLE,
  STATUS_TYPE,
  Option,
} from '@/types';
import {
  AVATAR_THUMBNAIL,
  DEFAULT_CHEMIST_DATA,
  ERROR_MESSAGE,
  FORM_VALIDATION_MESSAGE,
  MAX_IMAGE_SIZE,
  SUCCESS_MESSAGE,
  VALID_IMAGE_FORMAT,
} from '@/constants';

// Helper
import { clearErrorOnChange, getRoleIdByName } from '@/utils';
import { useToast } from '@/context/toast';
import { addUserToChemists } from '@/actions/auth';
import {
  addUser,
  getUserRoles,
  updatePublishUser,
  updateUser,
  uploadImageToImgbb,
} from '@/services';

// Rules
import { CHEMIST_FORM_VALIDATION } from './rule';

export type ChemistFormProps = {
  specialtyOptions: Option[];
  id?: string;
  data?: UserModel;
  onClose: () => void;
};

const ChemistForm = memo(
  ({
    id = '',
    data = DEFAULT_CHEMIST_DATA,
    specialtyOptions,
    onClose,
  }: ChemistFormProps) => {
    const {
      username = '',
      email = '',
      password = '',
      description = '',
      rating = 0,
      tasks = 0,
      reviews = 0,
      specialtyId,
      avatar,
    } = data || {};

    const { id: specialty } = specialtyId?.data || {};

    const [imageUpload, setImageUpload] = useState<string>('');
    const [formImage, setFormImage] = useState<FormData | null>(null);

    const [isPending, setIsPending] = useState(false);
    const [roles, getRoles] = useState<RolePermission[]>([]);

    const openToast = useToast();

    const hiddenFileInput = useRef<HTMLInputElement>(null);

    const {
      control,
      handleSubmit,
      getValues,
      clearErrors,
      setError: setFormError,
      formState: { errors, isValid, isLoading, isDirty },
      trigger,
    } = useForm<ChemistFormData>({
      mode: 'onBlur',
      reValidateMode: 'onBlur',
      defaultValues: {
        avatar: avatar || AVATAR_THUMBNAIL,
        username,
        password,
        confirmPassWord: password,
        email,
        description,
        specialtyId: specialty?.toString(),
      },
    });

    useEffect(() => {
      const fetchUserRoles = async () => {
        const { roles, error } = await getUserRoles();
        if (error) throw error;
        getRoles(roles);
      };

      fetchUserRoles();
    }, []);

    const isEdit = !!data.email;

    // Handle show hide password
    const [isShowPassword, setIsShowPassword] = useState<boolean>(false);
    const [isShowConfirmPassword, setIsShowConfirmPassword] =
      useState<boolean>(false);
    const handleToggleVisiblePassword = useCallback(
      () => setIsShowPassword((prev) => !prev),
      [],
    );
    const handleToggleShowConfirmPassword = useCallback(
      () => setIsShowConfirmPassword((prev) => !prev),
      [],
    );

    // Handle input file changes
    const handleUpload =
      (callback: (value: string) => void) =>
      (event: React.ChangeEvent<HTMLInputElement>) => {
        clearErrors('avatar');

        const files = event.target.files;
        if (files && files[0]) {
          const image = files[0];

          const isNoMoreThan32MB = image.size <= MAX_IMAGE_SIZE;

          if (!isNoMoreThan32MB) {
            setFormError('avatar', {
              message: FORM_VALIDATION_MESSAGE.MAX_SIZE({
                item: 'image',
                size: 32,
                unit: 'MB',
              }),
            });
            return false;
          }

          const hasValidFormat = VALID_IMAGE_FORMAT.includes(image.type);

          if (!hasValidFormat) {
            setFormError('avatar', {
              message: FORM_VALIDATION_MESSAGE.ACCEPTED_FORMATS,
            });
            return false;
          }

          const formData = new FormData();
          formData.append('image', image);

          setFormImage(formData);
          setImageUpload(URL.createObjectURL(image));
          callback(URL.createObjectURL(image));

          return true;
        }
      };

    // Handle remove upload image
    const handleRemoveImage = (callback: (value: string) => void) => () => {
      setImageUpload('');
      setFormImage(null);
      callback('');

      if (hiddenFileInput.current) {
        hiddenFileInput.current.value = '';
      }
    };

    const handleInputChange = useCallback(
      (name: keyof ChemistFormData, onChange: (value: string) => void) => {
        return (e: ChangeEvent<HTMLInputElement>) => {
          onChange(e.target.value);
          // Clear error message on change
          clearErrorOnChange(name, errors, clearErrors);
        };
      },
      [clearErrors, errors],
    );

    const handleError = useCallback(
      (error: string) => {
        if (error === ERROR_MESSAGE.DUPLICATE_FIELD)
          error = ERROR_MESSAGE.USERNAME;

        openToast({
          message: ERROR_MESSAGE.CREATE('chemist', error),
          type: STATUS_TYPE.ERROR,
        });
        setIsPending(false);
      },
      [openToast],
    );

    // Handle submit form data to create chemist
    const handleSubmitForm: SubmitHandler<ChemistFormData> = useCallback(
      async (formData) => {
        setIsPending(true);

        const { avatar, username, email, password, specialtyId, description } =
          formData;

        const avatarUpload =
          (formImage && (await uploadImageToImgbb(formImage)).image) || '';

        const payload: UserPayload = {
          username,
          email,
          description,
          specialtyId: Number(specialtyId),
          role: Number(getRoleIdByName(roles, ROLE.NORMAL_USER)),
          avatar: avatarUpload || avatar,
          ...(!isEdit && { password }),
        };

        if (isEdit) {
          const data = {
            ...payload,
            specialtyId: payload.specialtyId ? payload.specialtyId : null,
          };

          const { error } = await updateUser(id, data);

          if (error) {
            handleError(error);
            return;
          }

          openToast({
            message: SUCCESS_MESSAGE.UPDATE('chemist'),
            type: STATUS_TYPE.SUCCESS,
          });

          return;
        }

        const { user, error } = await addUser(payload);

        if (error) {
          handleError(error);
          return;
        }

        if (!user) return;

        const { id: userId = '' } = user;

        const updateUserResult = await updatePublishUser(userId);

        if (updateUserResult.error) {
          handleError(updateUserResult.error);
          return;
        }

        const addUserResult = await addUserToChemists({
          users_permissions_user: userId,
        });

        if (addUserResult.error) {
          handleError(addUserResult.error);
          return;
        }

        openToast({
          message: SUCCESS_MESSAGE.CREATE('chemist'),
          type: STATUS_TYPE.SUCCESS,
        });

        setIsPending(false);
        onClose();
      },
      [formImage, handleError, id, isEdit, onClose, openToast, roles],
    );

    const handleCloseSelect = (field: keyof ChemistFormData) => () => {
      trigger(field);
    };

    return (
      <form
        onSubmit={handleSubmit(handleSubmitForm)}
        className="p-4 flex flex-col"
        data-testid="chemist-form"
      >
        <Text variant="title" size="xl">
          {isEdit ? 'Update chemist' : 'Create chemist'}
        </Text>

        {/* Avatar */}
        <Controller
          control={control}
          name="avatar"
          render={({
            field: { value, onChange, ...rest },
            fieldState: { error },
          }) => (
            <>
              <ImageUpload
                {...rest}
                ref={hiddenFileInput}
                data-testid="chemist-avatar"
                src={value}
                srcUpload={imageUpload}
                onRemoveImage={handleRemoveImage(onChange)}
                onUploadImage={handleUpload(onChange)}
                isDisabled={isPending}
              />
              {!!error && (
                <p className="text-danger-100 text-xs ml-2 text-center">
                  {error.message}
                </p>
              )}
            </>
          )}
        />

        {/* Info section */}
        {isEdit && (
          <div className="flex justify-between self-center gap-6 my-4">
            <div className="flex items-center gap-1 sm:gap-2">
              <NoteIcon customClass="w-6 h-6" />
              <Text size="sm" variant="title" customClass="font-medium">
                {tasks} Tasks
              </Text>
            </div>

            <div className="flex items-center gap-1 sm:gap-2">
              <StarIcon customClass="w-6 h-6 text-light-orange" />
              <Text size="sm" variant="title" customClass="font-medium">
                {rating} ({reviews} Reviews)
              </Text>
            </div>
          </div>
        )}

        {/* Username */}
        <Controller
          name="username"
          control={control}
          render={({
            field: { name, onChange, ...rest },
            fieldState: { error },
          }) => (
            <Input
              {...rest}
              isRequired
              label="Username"
              labelPlacement="outside"
              name={name}
              placeholder="Please enter username"
              type="text"
              size="sm"
              isDisabled={isLoading || isEdit || isPending}
              isInvalid={!!error?.message}
              errorMessage={error?.message}
              onChange={handleInputChange(name, onChange)}
            />
          )}
          rules={CHEMIST_FORM_VALIDATION.USERNAME}
        />

        {/* Email */}
        <Controller
          name="email"
          control={control}
          render={({
            field: { name, onChange, ...rest },
            fieldState: { error },
          }) => (
            <Input
              {...rest}
              label="Email"
              isRequired
              labelPlacement="outside"
              name={name}
              placeholder="Please enter email"
              type="text"
              size="sm"
              isDisabled={isLoading || isEdit || isPending}
              isInvalid={!!error?.message}
              errorMessage={error?.message}
              onChange={handleInputChange(name, onChange)}
            />
          )}
          rules={CHEMIST_FORM_VALIDATION.EMAIL}
        />

        {!isEdit && (
          <>
            {/* Password */}
            <Controller
              name="password"
              control={control}
              render={({
                field: { onChange, name, ...rest },
                fieldState: { error },
              }) => (
                <Input
                  {...rest}
                  name={name}
                  label="Password"
                  labelPlacement="outside"
                  size="sm"
                  type={isShowPassword ? 'text' : 'password'}
                  placeholder="Please enter password"
                  endContent={
                    <Button
                      aria-label="visible password"
                      onClick={handleToggleVisiblePassword}
                      isIconOnly
                      className="p-0 min-w-5 h-5 text-primary-200"
                    >
                      {isShowPassword ? <EyeIcon /> : <EyeSlashIcon />}
                    </Button>
                  }
                  isDisabled={isLoading || isPending}
                  isInvalid={!!error?.message}
                  errorMessage={error?.message}
                  onChange={handleInputChange(name, onChange)}
                />
              )}
              rules={CHEMIST_FORM_VALIDATION.PASSWORD(
                getValues,
                setFormError,
                clearErrors,
                isEdit,
              )}
            />

            {/* Confirm Password */}
            <Controller
              name="confirmPassWord"
              control={control}
              render={({
                field: { onChange, name, ...rest },
                fieldState: { error },
              }) => (
                <Input
                  {...rest}
                  size="sm"
                  label="Confirm Password"
                  labelPlacement="outside"
                  type={isShowConfirmPassword ? 'text' : 'password'}
                  placeholder="Please enter confirm password"
                  endContent={
                    <Button
                      aria-label="visible confirm password"
                      onClick={handleToggleShowConfirmPassword}
                      isIconOnly
                      className="p-0 min-w-5 h-5 text-primary-200"
                    >
                      {isShowConfirmPassword ? <EyeIcon /> : <EyeSlashIcon />}
                    </Button>
                  }
                  isDisabled={isLoading || isPending}
                  isInvalid={!!error?.message}
                  errorMessage={error?.message}
                  onChange={handleInputChange(name, onChange)}
                />
              )}
              rules={CHEMIST_FORM_VALIDATION.CONFIRM_PASSWORD(
                getValues,
                isEdit,
              )}
            />
          </>
        )}

        {/* Specialty */}
        <Controller
          control={control}
          name="specialtyId"
          render={({
            field: { name, value, onChange, onBlur: _onBlur, ...rest },
            fieldState: { error },
          }) => (
            <Select
              {...rest}
              name={name}
              value={value}
              label="Specialty"
              data-testid="specialty-select"
              placeholder="Select specialty"
              labelPlacement="outside"
              variant="bordered"
              classNames={{
                mainWrapper: 'h-16',
                trigger: 'h-[auto] py-3 max-h-10',
                errorMessage: 'text-danger text-xs ml-2',
                label: 'top-5 text-sm text-foreground',
                value: 'text-sm text-primary-100',
              }}
              isDisabled={isLoading || isPending}
              selectedKeys={[value]}
              options={specialtyOptions}
              isInvalid={!!error?.message}
              errorMessage={error?.message}
              onChange={onChange}
              onClose={handleCloseSelect(name)}
            />
          )}
        />

        {/* Description */}
        <Controller
          name="description"
          control={control}
          render={({
            field: { name, value, onChange, ...rest },
            fieldState: { error },
          }) => (
            <Textarea
              {...rest}
              label="Description"
              labelPlacement="outside"
              name={name}
              value={value}
              size="sm"
              classNames={{
                inputWrapper:
                  'border-text-foreground border-1 data-[focus=true]:border-secondary-300 group-data-[focus-visible=true]:ring-0',
                errorMessage: 'text-danger text-xs ml-2',
                label: 'text-sm',
              }}
              isDisabled={isLoading || isPending}
              placeholder="Please enter description"
              isInvalid={!!error?.message}
              errorMessage={error?.message}
              onChange={handleInputChange(name, onChange)}
            />
          )}
          rules={CHEMIST_FORM_VALIDATION.DESCRIPTION}
        />

        <div className="h-[78px] flex flex-col justify-end">
          <div className="w-full gap-2 flex justify-end">
            <Button
              onClick={onClose}
              variant="outline"
              color="outline"
              className="font-medium"
              isDisabled={isPending}
            >
              Cancel
            </Button>
            <Button
              isDisabled={!isValid || !isDirty || isPending}
              isLoading={isLoading || isPending}
              type="submit"
              data-testid="submit"
            >
              Submit
            </Button>
          </div>
        </div>
      </form>
    );
  },
);

export default ChemistForm;
ChemistForm.displayName = 'ChemistForm';
