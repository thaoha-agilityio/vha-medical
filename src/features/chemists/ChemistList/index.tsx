// Components
import ChemistList from '@/features/chemists/ChemistList/ChemistList';

// Services
import { getChemists, getSpecialties } from '@/services';

// Types
import { MetaResponse } from '@/types';

interface ChemistsProps extends MetaResponse {
  searchParamsAPI: URLSearchParams;
  role: string;
  defaultSpecialty?: string;
}

const Chemists = async ({
  searchParamsAPI,
  role,
  defaultSpecialty,
}: ChemistsProps) => {
  const { specialties } = await getSpecialties({});
  const { chemists, pagination } = await getChemists({
    searchParams: searchParamsAPI,
  });

  return (
    <ChemistList
      chemists={chemists}
      pagination={pagination}
      defaultSpecialty={defaultSpecialty}
      role={role}
      specialties={specialties}
    />
  );
};

export default Chemists;
