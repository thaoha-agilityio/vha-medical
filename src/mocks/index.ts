export * from './appointments';
export * from './auth';
export * from './chemists';
export * from './common';
export * from './dateTime';
export * from './notification';
export * from './options';
export * from './profile';
export * from './table';
export * from './user';
