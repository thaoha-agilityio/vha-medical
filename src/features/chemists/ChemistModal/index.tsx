// Types
import { Option, UserModel } from '@/types';

// Components
import { BaseModal } from '@/components/ui/BaseModal';
import ChemistForm from '../ChemistForm';

export type ChemistModalProps = {
  isOpen: boolean;
  specialtyOptions: Option[];
  onClose: () => void;
  data?: UserModel;
  id?: string;
};

const ChemistModal = ({
  isOpen,
  onClose,
  specialtyOptions,
  id,
  data,
}: ChemistModalProps) => {
  return (
    <BaseModal isOpen={isOpen} onClose={onClose} placement="center" size="2xl">
      <ChemistForm
        data={data}
        id={id}
        onClose={onClose}
        specialtyOptions={specialtyOptions}
      />
    </BaseModal>
  );
};

export default ChemistModal;
ChemistModal.displayName = 'ChemistModal';
