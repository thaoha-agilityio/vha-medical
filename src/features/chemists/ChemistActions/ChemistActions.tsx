'use client';

import { useDisclosure } from '@nextui-org/react';
import { usePathname, useRouter, useSearchParams } from 'next/navigation';
import { ChangeEvent, memo, useCallback, useMemo, useState } from 'react';

// Components
import { Button, Select } from '@/components/ui';
import ChemistModal from '../ChemistModal';

// Types
import { APIResponse, MetaResponse, ROLE, SpecialtyModel } from '@/types';

// Utils
import { transformSpecialtiesById, transformSpecialtiesByName } from '@/utils';

export type ChemistActionsProps = {
  specialties: Array<APIResponse<SpecialtyModel>>;
  role: string;
  defaultSpecialty?: string;
} & MetaResponse;
const ChemistActions = ({
  specialties,
  role,
  defaultSpecialty = '',
}: ChemistActionsProps) => {
  const searchParams = useSearchParams() ?? '';
  const pathname = usePathname() ?? '';
  const { replace } = useRouter();

  const { isOpen, onOpen, onClose } = useDisclosure();

  const params = useMemo(
    () => new URLSearchParams(searchParams),
    [searchParams],
  );
  const specialtyOptionsById = transformSpecialtiesById(specialties);
  const specialtyOptionsByName = transformSpecialtiesByName(specialties);
  const [specialty, setSpecialty] = useState(
    new Set<string>([defaultSpecialty]),
  );

  const updateSearchParams = useCallback(
    (key: string, value: string) => {
      if (value) {
        params.set(key, value); // Update or add the key
      } else {
        params.delete(key); // Remove the key
      }

      replace(`${pathname}?${params}`);
    },
    [params, pathname, replace],
  );

  const handleSelectSpecialty = useCallback(
    (e: ChangeEvent<HTMLSelectElement>) => {
      const value = e.target.value;

      if (value !== specialty?.values().next().value) params.delete('page');

      if (value === 'all') {
        params.delete('specialty');
        replace(`${pathname}?${params}`);

        return;
      }

      const selectedSpecialty = specialtyOptionsByName.find(
        ({ key }) => key === value,
      );
      const { key = '' } = selectedSpecialty || {};

      setSpecialty(new Set([value]));
      updateSearchParams('specialty', key);
    },
    [
      params,
      pathname,
      replace,
      specialty,
      specialtyOptionsByName,
      updateSearchParams,
    ],
  );

  const isAdmin = role === ROLE.ADMIN;

  return (
    <div className="flex justify-between md:gap-4 mb-10 md:mb-0 ">
      <Select
        aria-label="Select Specialty"
        options={specialtyOptionsByName}
        selectedKeys={specialty}
        defaultSelectedKeys={specialtyOptionsByName[0].key}
        placeholder="Specialty"
        classNames={{
          innerWrapper: 'w-[180px]',
          trigger: 'w-[180px] h-[52px]',
          listbox: 'px-0',
        }}
        onChange={handleSelectSpecialty}
      />
      {isAdmin && (
        <Button className="font-medium h-[52px]" onClick={onOpen}>
          Create
        </Button>
      )}

      <ChemistModal
        isOpen={isOpen}
        onClose={onClose}
        specialtyOptions={specialtyOptionsById}
      />
    </div>
  );
};

export default memo(ChemistActions);
