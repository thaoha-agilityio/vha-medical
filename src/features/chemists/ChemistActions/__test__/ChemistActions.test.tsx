import { act, fireEvent, render, screen } from '@testing-library/react';

import { MOCK_SPECIALTIES } from '@/mocks/chemists';
import ChemistActions, { ChemistActionsProps } from '../ChemistActions';

const mockReplace = jest.fn();
jest.mock('next/navigation', () => ({
  useSearchParams: jest.fn().mockReturnValue(new URLSearchParams()),
  useRouter: () => ({ replace: mockReplace }),
  usePathname: jest.fn(),
}));

describe('ChemistActions', () => {
  const mockProps: ChemistActionsProps = {
    role: 'Admin',
    specialties: MOCK_SPECIALTIES,
  };

  const setup = async (props: ChemistActionsProps) =>
    act(() => render(<ChemistActions {...props} />));

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', async () => {
    const { asFragment } = await setup(mockProps);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should be able to filter by specialty', async () => {
    await setup(mockProps);

    const specialtyDropdown = screen.getByLabelText('Select Specialty');

    fireEvent.click(specialtyDropdown);

    const specialtyOption: HTMLLIElement = screen.getByRole('option', {
      name: /organic chemist/i,
    });

    fireEvent.click(specialtyOption);

    expect(specialtyOption.getAttribute('data-focus')).toBe('true');
  });
});
