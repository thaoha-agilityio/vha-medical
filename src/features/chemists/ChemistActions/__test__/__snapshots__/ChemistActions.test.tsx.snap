// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChemistActions should render correctly 1`] = `
<DocumentFragment>
  <div
    class="flex justify-between md:gap-4 mb-10 md:mb-0 "
  >
    <div
      class="group flex flex-col w-full transition-background motion-reduce:transition-none !duration-150 group relative justify-end data-[has-label=true]:mt-[calc(theme(fontSize.small)_+_10px)]"
      data-filled="true"
      data-slot="base"
    >
      <div
        aria-hidden="true"
        data-a11y-ignore="aria-hidden-focus"
        data-testid="hidden-select-container"
        style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
      >
        <input
          style="font-size: 16px;"
          tabindex="0"
          type="text"
          value=""
        />
        <label>
          <select
            size="6"
            tabindex="-1"
            title=""
          >
            <option />
            <option
              value="all"
            >
              All
            </option>
            <option
              value="instrumentation"
            >
              Instrumentation
            </option>
            <option
              value="laboratory_chemist"
            >
              Laboratory Chemist
            </option>
            <option
              value="organic_chemist"
            >
              Organic Chemist
            </option>
            <option
              value="power_plant_chemist"
            >
              Power Plant Chemist
            </option>
            <option
              value="qc_chemist"
            >
              QC Chemist
            </option>
          </select>
        </label>
      </div>
      <div
        class="w-full flex flex-col"
        data-slot="mainWrapper"
      >
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-label="Select Specialty"
          aria-labelledby="react-aria-:r8: react-aria-:r3:"
          class="relative gap-3 inline-flex flex-row items-center shadow-sm tap-highlight-transparent group-data-[focus=true]:bg-default-100 min-h-10 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-2.5 py-2 rounded-medium bg-background-100 shadow-stack border-1 outline-offset-0 data-[focus=true]:border-green data-[focus=true]:border-1 data-[focus=true]:bg-background-100 data-[hover=true]:bg-background-100 group-data-[invalid=true]:!border-danger-100 w-[180px] h-[52px]"
          data-slot="trigger"
          id="react-aria-:r3:"
          type="button"
        >
          <div
            class="inline-flex h-full min-h-4 items-center gap-1.5 box-border text-xs leading-[18px] group-data-[invalid=true]:!text-danger-100 w-[180px]"
            data-slot="innerWrapper"
          >
            <span
              class="w-full text-left rtl:text-right truncate group-data-[has-value=true]:text-default-foreground text-xs group-data-[invalid=true]:!text-danger-100 text-primary-100 font-medium"
              data-slot="value"
              id="react-aria-:r8:"
            >
              Specialty
            </span>
          </div>
          <svg
            aria-hidden="true"
            class="absolute end-3 w-4 h-4 transition-transform duration-150 ease motion-reduce:transition-none data-[open=true]:rotate-180 text-primary-100"
            data-slot="selectorIcon"
            fill="none"
            focusable="false"
            height="1em"
            role="presentation"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </div>
    </div>
    <button
      class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover border-transparent bg-linear-success text-content1 font-medium h-[52px]"
      type="button"
    >
      Create
    </button>
  </div>
</DocumentFragment>
`;
